### PR TITLE
Update oneDAL PCA infer method

### PR DESCRIPTION
**Pull request by md-shafiul-alam**
# Description
The changes are proposed to move some computations from python wrappers to oneDAL to improve performance of python libraries.

Changes proposed in this pull request:
- Oneapi model class for PCA has been extended to allow passing means, eigenvalues, and variances from python
- Support for whiten has been added in oneDAL if eigenvalues are passed
- Data for inference will be mean centered if mean is passed
- Three primitives have been added: Elementwise subtraction of 1d array from 2d array (required for mean centering), elementwise division of 2d array by 1d array (required for whitening), SQRT of an 1d array (required for whitening). The primitives are placed under stat temporarily.

### CODE REVIEW

File: cpp/oneapi/dal/algo/pca/backend/gpu/infer_kernel_dpc.cpp [hunk 1]
```diff
@@ -39,25 +39,65 @@ static result_t infer(const context_gpu& ctx, const descriptor_t& desc, const in
...
     dal::detail::check_mul_overflow(row_count, component_count);
 
     const auto data_nd = pr::table2ndarray<Float>(queue, data, sycl::usm::alloc::device);
+    auto means_nd = pr::table2ndarray_1d<Float>(queue, means, sycl::usm::alloc::device);
+    auto mean_centered_data_nd =
+        pr::ndarray<Float, 2>::empty(queue, { row_count, col_count }, sycl::usm::alloc::device);
     const auto eigenvectors_nd =
         pr::table2ndarray<Float>(queue, eigenvectors, sycl::usm::alloc::device);
 
+    pr::elementwise_difference(queue, row_count, data_nd, means_nd, mean_centered_data_nd);
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/pca/backend/gpu/infer_kernel_dpc.cpp:57]: I would suggest to use 
auto 'event_name' = pr::elementwise_difference(queue, row_count, data_nd, means_nd, mean_centered_data_nd);
and provide this event in gemm as deps

File: cpp/oneapi/dal/algo/pca/backend/gpu/infer_kernel_dpc.cpp [hunk 2]
```diff
@@ -39,25 +39,65 @@ static result_t infer(const context_gpu& ctx, const descriptor_t& desc, const in
     const auto data = input.get_data();
     auto model = input.get_model();
     auto eigenvectors = model.get_eigenvectors();
+    auto means = model.get_means();
+    auto eigenvalues = model.get_eigenvalues();
+
     const std::int64_t row_count = data.get_row_count();
+    const std::int64_t col_count = data.get_column_count();
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/pca/backend/gpu/infer_kernel_dpc.cpp:46]: I suggest to cover this code with additional checks e.g.
ONEDAL_ASSERT(data.get_column_count() > 0);

File: cpp/oneapi/dal/algo/pca/backend/gpu/infer_kernel_dpc.cpp [hunk 3]
```diff
@@ -39,25 +39,65 @@ static result_t infer(const context_gpu& ctx, const descriptor_t& desc, const in
...
+        const auto res_array_whitened = result_whitened_nd.flatten(queue, { whiten_event });
+        const auto res_table = homogen_table::wrap(res_array_whitened, row_count, component_count);
+
+        return result_t{}.set_transformed_data(res_table);
+    }
+
+    const auto res_array = res_nd.flatten(queue, { gemm_event });
+    const auto res_table = homogen_table::wrap(res_array, row_count, component_count);
     return result_t{}.set_transformed_data(res_table);
 }
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/pca/backend/gpu/infer_kernel_dpc.cpp:103]: it seems to me that our bazel tests almost do not cover infer for pca, might be a good time to add them

File: cpp/oneapi/dal/algo/pca/backend/gpu/infer_kernel_dpc.cpp [hunk 4]
```diff
@@ -39,25 +39,75 @@ static result_t infer(const context_gpu& ctx, const descriptor_t& desc, const in
...
         ONEDAL_PROFILER_TASK(gemm, queue);
-        gemm_event = pr::gemm(queue, data_nd, eigenvectors_nd.t(), res_nd, Float(1.0), Float(0.0));
+        gemm_event = pr::gemm(queue,
+                              mean_centered_data_nd,
+                              eigenvectors_nd.t(),
+                              res_nd,
+                              Float(1.0),
+                              Float(0.0),
+                              { mean_center_event });
+        gemm_event.wait_and_throw();
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/pca/backend/gpu/infer_kernel_dpc.cpp:80]: WHat is the point of waiting for the event here?

File: cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp [hunk 5]
```diff
@@ -51,6 +51,98 @@ sycl::event means(sycl::queue& q,
     });
 }
 
+// Subtract 1d array from 2d array elementwise
+template <typename Float>
+sycl::event elementwise_difference(sycl::queue& q,
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp:56]: Please use elementwise operations that are [already present](https://github.com/oneapi-src/oneDAL/blob/master/cpp/oneapi/dal/backend/primitives/element_wise.hpp)

File: cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp [hunk 6]
```diff
@@ -51,6 +51,98 @@ sycl::event means(sycl::queue& q,
...
+            const std::int64_t j = id[1];
+            difference_ptr[i * column_count + j] =
+                minuend_ptr[i * column_count + j] - subtrahend_ptr[j];
+        });
+    });
+}
+
+// SQRT of 1d array elementwise
+template <typename Float>
+sycl::event elementwise_sqrt(sycl::queue& q,
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp:91]: Same

File: cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp [hunk 7]
```diff
@@ -51,6 +51,98 @@ sycl::event means(sycl::queue& q,
...
+        cgh.depends_on(deps);
+        cgh.parallel_for(range, [=](sycl::id<1> id) {
+            dst_ptr[id[0]] = sycl::sqrt(src_ptr[id[0]]);
+        });
+    });
+}
+
+// Divide 2d array by 1d array elementwise
+template <typename Float>
+sycl::event elementwise_division(sycl::queue& q,
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp:112]: Same

**Comment by icfaust** [comment]
How does this PR compare to #2573 ?

**Comment by md-shafiul-alam** [comment]
> How does this PR compare to #2573 ?

2573 is an updated version of this. Closing this one as duplicate.

**Comment by md-shafiul-alam** [comment]
Changes have been moved to https://github.com/oneapi-src/oneDAL/pull/2573, closed as duplicate

**Pull request closed without merging**